### PR TITLE
world editor prototype

### PR DESCRIPTION
**Pull request by kostmo**
Towards #558
I was motivated to build this after finding that editing scenario maps directly in the YAML file is rather constraining.

## What I've implemented so far
* A small, collapsible panel to the left of the REPL containing World Editing status/operations.  Enter world-editing mode with CTRL+e to show the panel.
    * This works only in `--cheat` mode
* Terrain selection
    * A "picker"/"eye dropper" middle-click mechanism to select a terrain style to draw.
    * A pop-up selector to choose between the 5 different types of terrain.
* Drawing terrain with the left mouse button
* Saving a rectangular section of the world map (terrain only) to a file with CTRL+s
* Code organization
    * The complete state of the World Editor, including "painted overlays" of terrain, is contained within the `uiWorldEditor` field of `UIState` record.
    * The bulk of the World Editor functionality shall be in new modules
    * Some refactoring of `Controller.hs` and `View.hs` to extract functions utilized by the World Editor (towards #707)

## Vision

* The audience for this tooling is strictly envisioned to be Scenario authors.
    * Though, if we eventually allow swarm-lang to program the UI, there may be some common code to extract.
* The World Editor is intended to be compatible with a workflow of editing maps in text form within YAML scenario files.

# Demos
## Round-trip with random world

    stack run -- --scenario creative --seed 0 --cheat

Then Ctrl+e, tab down to the Save button, hit Enter to save the map
In another tab run:

    stack run -- --scenario mymap.yaml

Toggle between tabs to compare, observe the derived map is an identical 41x21 subset.

### CODE REVIEW

**Comment by xsebek** [comment]
@kostmo looks nice, but what do you mean by "unambiguous mapping in the scenario file"? A palette is a JSON object, so its keys are unique.

**Comment by kostmo** [comment]
I have found myself splitting some of the larger modules to get this PR to work.  Perhaps I should spread this work across two PRs: the first will have no World Editor code in it---it will just be moving pre-existing code into different modules (i.e. toward #707).  This should make the subsequent PR (adding World Editor functionality) easier to review.

If I do that, the splits on their own may not make a ton of sense unless viewed from the context of this current PR.  In other words, take a look at the "final" PR to see the direction I'm going with those splits.

File: src/Swarm/Game/Scenario/WorldDescription.hs [hunk 1]
```diff
@@ -53,9 +63,108 @@ instance FromJSONE (EntityMap, RobotMap) WorldDescription where
...
+
+instance ToJSON WorldDescriptionPaint where
+  toJSON w =
+    object
+      [ "default" .= defaultTerrain w
+      , "offset" .= offsetOrigin w
+      , "palette" .= Y.toJSON paletteKeymap
+      , "upperleft" .= ul w
+      , "map" .= Y.toJSON mapText
+      ]
```
Comment (kostmo) [inline src/Swarm/Game/Scenario/WorldDescription.hs:98]: Here's my custom JSON serializer

File: src/Swarm/Game/Scenario/EntityFacade.hs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import Control.Lens hiding (from, (.=), (<.>))
+import Data.Text (Text)
+import Data.Yaml as Y
+import Swarm.Game.Display (Display)
+import Swarm.Game.Entity qualified as E
+
+type EntityName = Text
+
+data EntityFacadeData = EntityFacadeData EntityName Display
```
Comment (xsebek) [inline src/Swarm/Game/Scenario/EntityFacade.hs:13]: Why not just name it `EntityFacade` as the module name suggests?

File: src/Swarm/Game/Scenario/Cell.hs [hunk 3]
```diff
@@ -20,15 +20,27 @@ import Swarm.Util.Yaml
...
   , cellRobots :: [IndexedTRobot]
   }
   deriving (Eq, Show)
 
+-- | A single cell in a world map, which contains a terrain value,
+--   and optionally an entity and robot.
+type Cell = ProtoCell Entity
+
+-- Note: This is used only for the purpose of WorldPalette
+instance ToJSON Cell where
```
Comment (xsebek) [inline src/Swarm/Game/Scenario/Cell.hs:35]: Since it appears naturally in the world palette it seems OK to me.

This is like having a warning on gardening tools that they are only supposed to be used safely in the garden.

File: src/Swarm/Game/Scenario/Cell.hs [hunk 4]
```diff
@@ -20,15 +20,27 @@ import Swarm.Util.Yaml
...
+  toJSON x =
+    toJSON $
+      catMaybes
```
Comment (xsebek) [inline src/Swarm/Game/Scenario/Cell.hs:38]: Why not:
```Haskell
toJSON = toJSON . fmap mkPaint
```

File: src/Swarm/Game/Scenario/WorldDescription.hs [hunk 5]
```diff
@@ -53,9 +63,108 @@ instance FromJSONE (EntityMap, RobotMap) WorldDescription where
...
+  M.fromList $ concatMap (map genTuple) cellGrid
+ where
+  genTuple c =
+    (toKey terrainEfd, terrainEfd)
+   where
+    terrainEfd = cellToTerrainEntityNamePair c
+
+-- | Note that display characters are not unique
+-- across different entities! However, the palette KeyMap
+-- as a conveyance serves to dedupe them.
```
Comment (xsebek) [inline src/Swarm/Game/Scenario/WorldDescription.hs:118]: Could this function be split to first create the updated palette?

File: src/Swarm/Game/Scenario/WorldDescription.hs [hunk 6]
```diff
@@ -53,9 +63,108 @@ instance FromJSONE (EntityMap, RobotMap) WorldDescription where
...
+  unassignedCells :: M.Map TerrainEntityNamePair TerrainEntityFacadePair
+  unassignedCells =
+    M.withoutKeys entityCells $
+      Set.fromList $
+        map (toKey . snd) preassignments
+
+  unassignedCharacters :: Set.Set Char
+  unassignedCharacters =
+    -- TODO: Make this pool larger?
+    Set.difference (Set.fromList ['a' .. 'z']) $
```
Comment (xsebek) [inline src/Swarm/Game/Scenario/WorldDescription.hs:144]: I would suggest filtering all the `Char` values based on their Unicode category.

Letters, numbers, symbols and marks should give us a big enough pool. 🙂

File: src/Swarm/Game/Scenario/EntityFacade.hs (hunk 2 above)
Comment (xsebek) [reply src/Swarm/Game/Scenario/EntityFacade.hs:13]: Also, do you need the whole `Display`? I would expect only needing a `Char`.

**Review by xsebek — COMMENTED** [review]
Cool, I think I understand the reasoning better after reading through this PR. 👍 

Can't wait to use it to create new scenarios. 🎨 👨‍🎨

File: src/Swarm/Game/Scenario/WorldDescription.hs (hunk 6 above)
Comment (kostmo) [reply src/Swarm/Game/Scenario/WorldDescription.hs:144]: Yes, I was thinking about that.  And `Data.Char` has functions that allow filtering by category.  But what should the code look like that generates the initial list to be filtered?

File: src/Swarm/TUI/Editor/EditorView.hs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    cornerCoords <- case we ^. editingBounds . boundsSelectionStep of
+      LowerRightPending cornerCoords -> Just cornerCoords
+      _ -> Nothing
+    pure $ EU.isOutsideTopLeftCorner cornerCoords coords
+
+drawLabeledEntitySwatch :: EntityFacadeData -> Widget Name
+drawLabeledEntitySwatch (EntityFacadeData eName eDisplay) =
+  tile <+> txt eName
+ where
+  tile = padRight (Pad 1) $ renderDisplay eDisplay
```
Comment (kostmo) [inline src/Swarm/TUI/Editor/EditorView.hs:148]: Here's where I use the whole `Display` rather than just the `Char`

File: src/Swarm/Game/Scenario/EntityFacade.hs (hunk 2 above)
Comment (kostmo) [reply src/Swarm/Game/Scenario/EntityFacade.hs:13]: Turns out I need to use the `Display` [here](https://github.com/swarm-game/swarm/pull/873/files#r1053872827).

File: src/Swarm/TUI/Model/StateUpdate.hs [hunk 8]
```diff
@@ -137,3 +145,20 @@ scenarioToUIState siPair u = do
...
+ where
+  myWorld = scenario ^. scenarioWorld
+  Location left top = ul myWorld
+
+  setNewBounds maybeOldBounds =
+    if mapHeight /= 0 && mapWidth /= 0
+      then Just newBounds
+      else maybeOldBounds
+
+  -- TODO Note inversion of horizontal and vertical coordinates
```
Comment (kostmo) [inline src/Swarm/TUI/Model/StateUpdate.hs:159]: use `locToCoords` or `coordsToLoc` here

File: src/Swarm/TUI/Editor/EditorView.hs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        [ str "Area:"
+        , areaContent
+        ]
+
+  renderBounds (W.Coords (x1, y1), W.Coords (x2, y2)) =
+    str $
+      L.intercalate
+        " @ "
+        [ rectSize
+        , show (y1, x1) -- Note the inverted coords!
```
Comment (kostmo) [inline src/Swarm/TUI/Editor/EditorView.hs:100]: use `locToCoords` or `coordsToLoc`

File: src/Swarm/TUI/Editor/Util.hs [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  Maybe (W.Coords, W.Coords) ->
+  W.World Int Entity ->
+  [[CellPaintDisplay]]
+getEditedMapRectangle _ Nothing _ = []
+getEditedMapRectangle worldEditor (Just (tl, br)) w =
+  map renderRow [yTop .. yBottom]
+ where
+  -- We swap the horizontal and vertical coordinate.
+  -- TODO If this is necessary, then what is mouseLocToWorldCoords for?
+  toWorldCoords (W.Coords (mx, my)) = W.Coords (my, mx)
```
Comment (kostmo) [inline src/Swarm/TUI/Editor/Util.hs:79]: Replace this with `locToCoords` or `coordsToLoc`

File: src/Swarm/TUI/View.hs [hunk 11]
```diff
@@ -843,51 +862,7 @@ drawWorld showRobots g =
       let w = ctx ^. availWidthL
           h = ctx ^. availHeightL
           ixs = range (viewingRegion g (fromIntegral w, fromIntegral h))
-      render . vBox . map hBox . chunksOf w . map (drawLoc showRobots g) $ ixs
-
--- | Render the 'Display' for a specific location.
-drawLoc :: Bool -> GameState -> W.Coords -> Widget Name
```
Comment (kostmo) [inline src/Swarm/TUI/View.hs:849]: Move these functions in another PR

File: src/Swarm/Game/Scenario/Cell.hs (hunk 4 above)
Comment (kostmo) [reply src/Swarm/Game/Scenario/Cell.hs:45]: hmm, that doesn't typecheck?

**Comment by byorgey** [comment]
How's this going?  Did you use it to create the map in #1136 ?

**Comment by kostmo** [comment]
> How's this going? Did you use it to create the map in #1136 ?

I actually did #1136 the traditional way.  I can't claim that the world editor is super useful in its current state.

However, although the feature is incomplete, I believe there is value in merging it as-is.  It is non-invasive to the game UI and it makes for a decent foundation to resume work on a better editor UX when time allows.  And it has a bunch of utility functions that I could imagine re-using for other features.

Keeping on top of the merge conflicts from `main` is a chore, and if #1138 were developed in parallel it will be a major effort to keep rebasing.

**Comment by byorgey** [comment]
@kostmo sorry it's taken me a while to get to it, but I'm now slowly making my way through this PR.  I have some suggestions, but ultimately I think merging it as a feature toggle makes a lot of sense.

**Review by byorgey — COMMENTED** [review]
Sorry it's taken me so long to look at this.  In general I agree that we should try to merge this, perhaps at first in such a way that it's disabled by default.  I think my biggest meta-comment is that this is a lot of code with very few comments, which made it difficult to review.  I lost steam somewhere in `TUI.Editor.Controller` --- it was a lot of work to try to figure out what all the code was for.  I'm not so much looking for comments that explain what the code *does*, but comments explaining things like purpose, design choices, where the code fits into the big picture, etc.

File: src/Swarm/TUI/Controller.hs [hunk 12]
```diff
@@ -421,6 +434,14 @@ handleModalEvent = \case
     Brick.zoom (uiState . uiModal . _Just . modalDialog) (handleDialogEvent ev)
     modal <- preuse $ uiState . uiModal . _Just . modalType
     case modal of
+      Just TerrainPaletteModal -> do
+        lw <- use $ uiState . uiWorldEditor . terrainList
+        newList <- refreshList lw
+        uiState . uiWorldEditor . terrainList .= newList
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/Swarm/TUI/Controller.hs:440]: Simplified with `Brick.zoom`

File: src/Swarm/Game/Scenario/WorldDescription.hs [hunk 13]
```diff
@@ -66,3 +74,112 @@ paintMap pal = traverse (traverse toCell . into @String) . T.lines
   toCell c = case KeyMap.lookup (Key.fromString [c]) (unPalette pal) of
     Nothing -> fail $ "Char not in world palette: " ++ show c
     Just cell -> return cell
+
+------------------------------------------------------------
+-- World editor
+------------------------------------------------------------
+
+type TerrainEntityNamePair = (TerrainType, Maybe EntityName)
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/Swarm/Game/Scenario/WorldDescription.hs:82]: Replaced with a parameterized type alias `TerrainWith a`

File: src/Swarm/Game/Scenario/WorldDescription.hs [hunk 14]
```diff
@@ -66,3 +74,112 @@ paintMap pal = traverse (traverse toCell . into @String) . T.lines
...
+      ]
+   where
+    cellGrid = area w
+    suggestedPalette = palette w
+    (mapText, paletteKeymap) = prepForJson suggestedPalette cellGrid
+
+toKey :: TerrainEntityFacadePair -> TerrainEntityNamePair
+toKey = fmap $ fmap (\(EntityFacade eName _display) -> eName)
+
+getUniquePairs :: [[CellPaintDisplay]] -> M.Map TerrainEntityNamePair TerrainEntityFacadePair
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/Swarm/Game/Scenario/WorldDescription.hs]: I have now added a comment to this (renamed) function.  Is this rationale OK?  Maybe there's a better way:
> -- | We want to identify all of the unique (terrain, entity facade) pairs.
-- However, "EntityFacade" includes a "Display" record, which contains more
-- fields than desirable for use as a unique key.
-- Therefore, we extract just the entity name for use in a
-- (terrain, entity name) key, and couple it with the original
-- (terrain, entity facade) pair in a Map.

File: src/Swarm/Game/Scenario/Cell.hs [hunk 15]
```diff
@@ -62,3 +74,19 @@ instance FromJSONE (EntityMap, RobotMap) Cell where
     robs <- mapMaybeM name2rob (drop 2 tup)
 
     return $ Cell terr ent robs
+
+------------------------------------------------------------
+-- World editor
+------------------------------------------------------------
+
+type CellPaintDisplay = PCell EntityFacade
```
[the remark this answers is not in the document]
Comment (kostmo) [reply src/Swarm/Game/Scenario/Cell.hs:88]: Added comments

**Review by byorgey — APPROVED** [review]
:rocket:
